[ios] set minimum ios version to 11.0, fix deprecations and cleanup dependencies

phunkyfish opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

Set minimum supported iOS version of 11.0. Removes support for 32bit iOS devices (which covers iPad 1-4 and iPhone 5 and earlier).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

Simplifies patching a logic around the iOS platform, especially around addons.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [X] **None of the above** (please explain below)

Not quite sure how to categorise this.

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [X] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

phunkyfish commented:
Early days for this one. Plenty more to do.

kambala-decapitator commented:
- tools/depends/target/fontconfig/fix-aarch64_atomics.patch:17 (if this is Kodi's patch)
- xbmc/platform/darwin/ios/XBMCController.mm:710

phunkyfish commented:
@kambala-decapitator the first is still required for 64bit if I'm not mistaken. For the second itis already set to 11.0 which is what it should be, correct?

kambala-decapitator commented:
oh right, first is patch :)

in the second you should remove the condition, as it's always true now.

phunkyfish commented:
I assume I can condense it to this:
`return UIEdgeInsetsInsetRect(self.view.bounds, m_window.safeAreaInsets);`

phunkyfish commented:
@Rechi should removing ios from Jenkins be part of the PR or would that be something that must be done manually on/before merge?

fuzzard commented:
https://github.com/xbmc/xbmc/blob/a353ae2276a647f4f30f7fcf86baac6afc124d6c/tools/depends/configure.ac#L360-L372

Can be rewritten to something like 

```
        use_cpu=arm64
        platform_min_version="$target_platform-version-min=11.0"
        if test "$use_platform" = "tvos"; then
          target_platform=appletvos
        else
          target_platform=iphoneos
        fi
```

fuzzard commented:
https://github.com/xbmc/xbmc/blob/a353ae2276a647f4f30f7fcf86baac6afc124d6c/tools/depends/configure.ac#L378-L396

Can be rewritten like 

```
        case $use_sdk in
          11.*);;
          12.*);;
          13.*);;
            *)
              AC_MSG_ERROR(error in configure of --with-sdk=$use_sdk)
            ;;
        esac

```

fuzzard commented:
https://github.com/xbmc/xbmc/blob/a353ae2276a647f4f30f7fcf86baac6afc124d6c/tools/depends/configure.ac#L400-L403

can be removed

fuzzard commented:
There is a line in README.iOS.md under section 2. Prequisites

```
Device with iOS 9.0 or newer to install Kodi after build.
```

Will need updating.

fuzzard commented:
End of section 4 in the same readme, theres a NOTE that references an ios 9 build.

```
NOTE: Advanced developers may want to specify an iOS SDK version (if multiple versions are installed) in the configure line(s) shown above. The example below would use the iOS SDK 9.0:

./configure --host=arm-apple-darwin --with-cpu=arm64 --with-sdk=9.0
```

Probably need to just alter that to 11

fuzzard commented:
https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/xbmc/platform/darwin/DarwinUtils.mm#L178-L179

should be removable. The line below it should most likely stay as i believe its relevant for ios10/11/12

fuzzard commented:
https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/buildsteps/ios/package#L10-L15

no need for the check now with a single ios platform

```
UPLOAD_FILENAME="kodi-$(getBuildRevDateStr)-ios64.deb"
```

fuzzard commented:
https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/darwin/packaging/darwin_embedded/mkdeb-darwin_embedded.sh.in#L31-L34

No need for the check, just set ARM64

fuzzard commented:
Can probably remove the previous ARM64 check completely, and then

https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/darwin/packaging/darwin_embedded/mkdeb-darwin_embedded.sh.in#L69-L70

Remove the $ARM64 && as every package will be 64bit

fuzzard commented:
https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/darwin/packaging/darwin_embedded/mkdeb-darwin_embedded.sh.in#L80-L89

Change to just

```
echo "Package: $PACKAGE_ARM64"                  >  $DIRNAME/$PACKAGE/DEBIAN/control
echo "Name: @APP_NAME@-${PP_DEVICE} (64-bit)"   >> $DIRNAME/$PACKAGE/DEBIAN/control
echo "Pre-Depends: cy+cpu.arm64"                >> $DIRNAME/$PACKAGE/DEBIAN/control
echo "Conflicts: $PACKAGE"                      >> $DIRNAME/$PACKAGE/DEBIAN/control
echo "Replaces: $PACKAGE"                       >> $DIRNAME/$PACKAGE/DEBIAN/control
```

fuzzard commented:
Remove

https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/darwin/Support/Codesign.command#L10

This can be reduced to just the LDID64 lines

https://github.com/xbmc/xbmc/blob/d14495a880c14a1c703a26ab96120108d514feee/tools/darwin/Support/Codesign.command#L14-L19

Rechi commented:
First get this PR ready.
Once it is approved adjust the Jenkins configuration, run another PR build and if that reports success merge the PR.

Rechi commented on docs/README.iOS.md:
```diff
@@ -75,25 +76,17 @@ git clone https://github.com/xbmc/xbmc kodi
 
 **TIP:** Look for comments starting with `Or ...` and only execute the command(s) you need.
 
 Configure build for 64bit (**recommended**):
```
What is the not recommended alternative?

phunkyfish replied:
Fair point. Removed the recommended.

Rechi commented on tools/buildsteps/defaultenv:
```diff
@@ -22,7 +22,7 @@ case $XBMC_PLATFORM_DIR in
     DEFAULT_XBMC_DEPENDS_ROOT=$WORKSPACE/tools/depends/xbmc-depends
     DEFAULT_CONFIGURATION="Debug"
-    DEFAULT_DARWIN_ARM_CPU="armv7"
+    DEFAULT_DARWIN_ARM_CPU="arm64"
```
Is there any case where the darwin arm cpu isn't arm64?

Rechi commented on tools/darwin/packaging/darwin_embedded/mkdeb-darwin_embedded.sh.in:
```diff
@@ -9,7 +9,7 @@ SWITCH=`echo $1 | tr [A-Z] [a-z]`
 DSYM_TARGET_DIR=/Users/Shared/xbmc-depends/dSyms
 DSYM_FILENAME=@APP_NAME@.app.dSYM
-ARM64=false
+ARM64=true
```
Is there any case where ARM64 isn't true?

phunkyfish replied:
I don't believe so

Rechi commented on tools/depends/configure.ac:
```diff
@@ -357,52 +357,29 @@ case $host in
         MC_CHECK_NOT_CPU([$use_cpu], "*86")
 
         # setup which sdk to use
+        use_cpu=arm64
```
Keep the possibility to override the cpu.

Rechi commented on docs/README.iOS.md:
````diff
@@ -103,9 +96,9 @@ make -j$(getconf _NPROCESSORS_ONLN)
+**NOTE:** **Advanced developers** may want to specify an iOS SDK version (if multiple versions are installed) in the configure line(s) shown above. The example below would use the iOS SDK 11.0:
 ```
-./configure --host=arm-apple-darwin --with-cpu=arm64 --with-sdk=9.0
+./configure --host=arm-apple-darwin --with-cpu=arm64 --with-sdk=11.0
````
Why pass `--with-cpu=arm64`?

Rechi commented on tools/depends/target/ffmpeg/CMakeLists.txt:
```diff
@@ -64,9 +64,7 @@ elseif(CORE_SYSTEM_NAME STREQUAL android)
-  if(NOT CPU MATCHES arm64)
-    list(APPEND ffmpeg_conf --cpu=cortex-a8)
-  else()
+  if(CPU MATCHES arm64)
```
Is there any case where CPU isn't arm64?

phunkyfish replied:
I don't believe so, but if there is then I should revert this?

Rechi commented on tools/depends/target/ffmpeg/Makefile:
```diff
@@ -44,10 +44,7 @@ ifeq ($(OS), android)
-    ffmpg_config += --cpu=cortex-a8
-    ffmpg_config += --as="$(NATIVEPREFIX)/bin/gas-preprocessor.pl -- $(CC)"
-  else
+  ifeq ($(CPU), arm64)
```
Is there any case where CPU isn't arm64?

Rechi reviewed (changes requested):
- ldid for 32 bit is still downloaded (extracted and installed)
- deployment target isn't updated
- There is still a dependency which explicitly disables a function, which is now available.

phunkyfish committed:
[ios] set minimum ios version to 11.0 and cleanup dependencies

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -7,23 +7,16 @@ DEPS= ../../Makefile.include Makefile
 LIBNAME=ldid
 VERSION32=1.0.0
```
unused?

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -7,23 +7,16 @@ DEPS= ../../Makefile.include Makefile
 LIBNAME=ldid
 VERSION32=1.0.0
 VERSION64=2
```
Why 64?

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -7,23 +7,16 @@ DEPS= ../../Makefile.include Makefile
 VERSION32=1.0.0
 VERSION64=2
-SOURCE32=$(LIBNAME)-$(VERSION32)
 SOURCE64=$(LIBNAME)-$(VERSION64)
```
Why 64?

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -7,23 +7,16 @@ DEPS= ../../Makefile.include Makefile
-SOURCE32=$(LIBNAME)-$(VERSION32)
 SOURCE64=$(LIBNAME)-$(VERSION64)
-ARCHIVE32=$(SOURCE32).tar.gz
 ARCHIVE64=$(SOURCE64).tar.gz
```
Why 64?

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -7,23 +7,16 @@ DEPS= ../../Makefile.include Makefile
-	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE32)
-	mv $(PLATFORM)/ldid $(PLATFORM)/ldid32
 	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE64)
 	mv $(PLATFORM)/ldid $(PLATFORM)/ldid64
```
Why 64?

phunkyfish replied:
Fixed, also updated command call in codesign command.

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -5,27 +5,19 @@ DEPS= ../../Makefile.include Makefile
-	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE64)
-	mv $(PLATFORM)/ldid $(PLATFORM)/ldid64
+	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
+	mv $(PLATFORM)/ldid $(PLATFORM)/ldid
```
Why rename to the same name?

phunkyfish replied:
Done

Rechi commented on tools/depends/native/ldid/Makefile:
```diff
@@ -5,27 +5,19 @@ DEPS= ../../Makefile.include Makefile
+	mv $(PLATFORM)/ldid $(PLATFORM)/ldid
 
 .installed-$(PLATFORM):$(PLATFORM)
 	cp $(PLATFORM)/ldid* $(NATIVEPREFIX)/bin
```
`ldid*`?

phunkyfish replied:
Done

phunkyfish commented:
@Rechi should be good now.

Rechi commented on xbmc/platform/darwin/ios/Info.plist.in:
```diff
@@ -119,7 +119,7 @@
 	<key>UIRequiredDeviceCapabilities</key>
 	<array>
-		<string>armv7</string>
+		<string>armv64</string>
```
Are you sure?

phunkyfish replied:
No I’m not

phunkyfish replied:
I am now: https://developer.apple.com/documentation/bundleresources/information_property_list/uirequireddevicecapabilities

phunkyfish replied:
Whoops, extra `v`

Rechi reviewed (changes requested):
- Jenkins depends configuration differs from what the readme tells.
- generated DEBIAN/control file looks wrong
- There is a patch hunk which worked around a function, which is now available.
- 1 new warning
- host is now actually `aarch64-apple-darwin` and not `arm-apple-darwin`

phunkyfish commented:
> * Jenkins depends configuration differs from what the readme tells.
> * generated DEBIAN/control file looks wrong
> * There is a patch hunk which worked around a function, which is now available.
> * 1 new warning
> * host is now actually `aarch64-apple-darwin` and not `arm-apple-darwin`

Ok, I fixed the warning and renamed armv64 to arm64 in the plist, also set the default darwin arm cpu correctly.

I'll need your help with these as I'm unsure what they should be:

* generated DEBIAN/control file looks wrong. Please help here.
* There is a patch hunk which worked around a function, which is now available. I've searching but i can't find this.
* host is now actually `aarch64-apple-darwin` and not `arm-apple-darwin`. Is arm-apple-darwin what it should be?

Rechi commented on tools/buildsteps/defaultenv:
```diff
@@ -4,7 +4,7 @@ NDK_VERSION=${NDK_VERSION:-"Default"}
 Configuration=${Configuration:-"Default"}
 XBMC_DEPENDS_ROOT=${XBMC_DEPENDS_ROOT:-"Default"}
-DARWIN_ARM_CPU=${DARWIN_ARM_CPU:-"Default"}
+DARWIN_ARM_CPU=${DARWIN_ARM_CPU:-"arm64"}
```
Is there any case where the darwin arm cpu isn't arm64?

fuzzard replied:
Not currently

kambala-decapitator replied:
there exists `arm64e`, but it can run `arm64` code

phunkyfish replied:
I understand something should be changed because it's the only option I just don't know what you would like changed.

Rechi replied:
Replace the usage of `DARWIN_ARM_CPU` with it's actual value.

Rechi commented:
- DEBIAN/control
```sh
echo "Package: $PACKAGE"                  >  $DIRNAME/$PACKAGE/DEBIAN/control
# [...]
echo "Conflicts: $PACKAGE"                      >> $DIRNAME/$PACKAGE/DEBIAN/control
echo "Replaces: $PACKAGE"                       >> $DIRNAME/$PACKAGE/DEBIAN/control
```
- `git grep -i clock_gettime -- tools/depends/`
- Currently one configures with `--host=arm-apple-darwin`, but the result are actually binaries for `aarch64-apple-darwin`.

phunkyfish commented:
Thanks for patch hint. Found it now.

phunkyfish commented:
Ok, just these left:

- tools/buildsteps/defaultenv - arm64
- Currently one configures with `--host=arm-apple-darwin`, but the result are actually binaries for `aarch64-apple-darwin`.

Rechi commented:
The resulting binaries are correct, they are the same as before. But arm CPU is arm for 32 bit and aarch64 CPU is arm for 64 bit.

phunkyfish commented:
So do you mean we just need to  update the  docs so the host is specified as this?

Would this not be the same for tvOS?

Rechi commented on tools/buildsteps/ios/configure-depends:
```diff
@@ -8,6 +8,6 @@ then
     --host=arm-apple-darwin \
     --with-sdk=$SDK_VERSION \
-    --with-cpu=$DARWIN_ARM_CPU \
+    --with-cpu=arm64 \
```
Why does this configure in a different way then the documentation tells?

Rechi commented:
Updating the documentation is only a part of it. Yes it's the same for tvOS.

phunkyfish commented:
So cpu should be set to aarch64 across the board for darwin (iOS/tvOS)? arm64 is just incorrect.

Rechi commented:
The host triple `arm-apple-darwin`, was always wrong for arm64/aarch64 builds. As now there are only those kind of builds, the host triple should be reflect that.
You'll also find hacks, because of the wrong host triple usage.

phunkyfish commented:
Ok, so if I do that as part of this PR it will mean also making tvOS changes. That’s ok to do?

phunkyfish commented:
So a question. Where exactly should it be set to arm64 and where to aarch64?

Or is just aarch64 everywhere instead of arm/arm64?

Rechi commented:
You can do the changes in this PR. Preferred is to have a commit fixing the host triple name before the minimum version bump, but you can combine it into the minimum version bump commit.
The CPU part of the host triplet has to be aarch64. Everywhere else you can keep arm64.

phunkyfish commented:
@Rechi - I may need your help sourcing the whereabouts of the hacks. I only found one on the host triplet but clearly there are many more.

Rechi commented:
I only found one more, `git grep aarch64-apple-darwin`.

phunkyfish commented:
Hmmm, strange. That's the command I used but I didn't find another one with it.

phunkyfish commented:
Anyone have any idea? After changing the host triplet in the second commit both ios64 and tvOS both fail on gnutls.

```
09:57:53 make[8]: *** No rule to make target `macosx/sha1-armv8.s', needed by `macosx/sha1-armv8.lo'.  Stop.
09:57:53 make[8]: *** Waiting for unfinished jobs....
```

fuzzard commented:
check slack for a diff to work around the fault.

Rechi commented on tools/depends/target/gnutls/Makefile:
```diff
@@ -28,9 +28,9 @@ $(TARBALLS_LOCATION)/$(ARCHIVE):
-	cd $(PLATFORM); $(ARCHIVE_TOOL) $(ARCHIVE_TOOL_FLAGS) $(TARBALLS_LOCATION)/$(ARCHIVE)
 	cd $(PLATFORM); patch -p1 -i ../size-max.patch
 	cd $(PLATFORM); patch -p1 -i ../add-dl-as-private-lib.patch
+	cd $(PLATFORM); patch -p1 -i ../disable-hw-accel-aarch64.patch
```
Why disable hardware acceleration for all aarch64 cpus?

phunkyfish replied:
Good point

Rechi commented on tools/depends/target/config.site.in:
```diff
@@ -106,6 +95,10 @@ fi
   ac_cv_func_sigaltstack=no
+  if test "${PACKAGE_NAME}" = "gnutls"; then
+    ac_cv_func_vfork_works=no
+    ac_cv_func_fork=no
```
Does `fork` and `vfork` work in general on appletvos?

phunkyfish replied:
No, it appears not.

Rechi replied:
In that case, why is it only disabled for a package named gnutls?

Rechi commented on tools/depends/target/gnutls/Makefile:
```diff
@@ -1,23 +1,16 @@
 
-ifeq (darwin, $(findstring darwin, $(HOST)))
-# darwins tar doesn't know about xz - so we need our native version of tar here
-ARCHIVE_TOOL=$(ARCHIVE_TOOL_NATIVE)
```
How is this change related to fixing the host triplet to `aarch64-apple-darwin`?

phunkyfish replied:
Moving to separate commit

Rechi replied:
Isn't this variable (`ARCHIVE_TOOL_NATIVE`) now unused?

phunkyfish replied:
So remove tar altogether from native depends?

phunkyfish replied:
Or is it needed to other stuff?

Rechi replied:
If it is unused or not required, yes.

phunkyfish replied:
tar needed for dpkg on ios/tvos so it stays I guess.

https://github.com/xbmc/xbmc/blob/1eff5afda088f8fc58b63f8a7313c2b71016b3e2/tools/depends/native/Makefile#L16-L18

Rechi replied:
Is it really needed by dpkg, or did only previously system tar versions miss some features?

phunkyfish replied:
Let’s remove it and see

phunkyfish replied:
Fails on darwin_embedded with it removed. Must be required.

phunkyfish replied:
Hmm, or not. I forgot to remove the dependency.

phunkyfish replied:
tar/xz still needed for ios/tvos.

phunkyfish commented:
jenkins build this please

phunkyfish commented:
jenkins build this please

phunkyfish commented:
jenkins build this please

phunkyfish commented:
@Rechi, this should be everything now i think

Rechi commented on tools/depends/target/gnutls/Makefile:
```diff
@@ -31,6 +24,9 @@ $(PLATFORM): $(TARBALLS_LOCATION)/$(ARCHIVE) $(DEPS)
 	cd $(PLATFORM); patch -p1 -i ../size-max.patch
 	cd $(PLATFORM); patch -p1 -i ../add-dl-as-private-lib.patch
+ifeq ($(OS),darwin_embedded)
+	cd $(PLATFORM); patch -p1 -i ../disable-hw-accel-aarch64.patch
```
Why patch the buildsystem, instead of usinging the already existing configuration switch?

phunkyfish replied:
I didn't realise there was a config switch. Will change that tomorrow.

phunkyfish committed:
[darwin_embedded] Move to correct aarch64-apple-darwin host triplet for tvOS/iOS

phunkyfish committed:
[apple] tar on osx now supports xz files no need for our native one

phunkyfish commented:
Jenkins build this please

phunkyfish commented:
Ok, @Rechi, think it's ready now for the jenkins changes now.

Rechi commented on tools/depends/target/config.site.in:
```diff
@@ -103,9 +92,12 @@ if test "@platform_os@" = "darwin_embedded"; then
 if test "@target_platform@" = "appletvos"; then
+  # Python3 build
   ac_cv_func_sigaltstack=no
+  # gnutls build
```
Why add the lib info? This info might get outdated if libraries are updated.
What useful information does this comment provide?

phunkyfish replied:
Easy to remove.

Rechi commented on tools/depends/target/config.site.in:
```diff
@@ -103,9 +92,12 @@ if test "@platform_os@" = "darwin_embedded"; then
   ac_cv_func_sigaltstack=no
+  # gnutls build
+  ac_cv_func_vfork_works=no
+  ac_cv_func_fork=no
```
Doesn't that apply for binary add-ons too?

phunkyfish replied:
You mean also set this in `config-binaddons.site.in`. It wasn't required before now.

phunkyfish commented:
jenkins build this please

phunkyfish commented:
Ok, ready @Rechi

phunkyfish commented:
@Rechi is it ok to make the Jenkins changes now?

lrusak commented:
> There is still a (new) hack because of the wrong host triplet.

Could you be more specific? Do you mean about the mixed usage of `arm64` and `aarch64`?

phunkyfish commented:
Maybe I’ve been looking at this too long but I’m not seeing it.

Rechi commented:
> This branch is 3 commits ahead, 113 commits behind xbmc:master.

https://github.com/phunkyfish/xbmc/tree/min-ios-version

So there are changes from 113 commits which are currently not included in your branch.

phunkyfish commented:
Ah, that would make sense why I’m not seeing it so. At least I didn’t add the new hack!

phunkyfish commented:
Ok, I removed the todo's in the two toolchain files. Any other hacks you came across @Rechi ?

I didn't spot anything additional.

phunkyfish commented:
jenkins build this please

phunkyfish commented:
@Rechi is it ok to make the jenkins changes now?

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please
